src/infra/flow-system/postgres/src/postgres_flow_event_store.rs

```diff
@@ -0,0 +1,651 @@
+                    .map_err(|e| sqlx::Error::Decode(Box::new(e)))?;
+
+                Ok((EventID::new(event_row.event_id), event))
+            })
+            .fetch(connection_mut)
+            .map_err(|e| GetEventsError::Internal(e.int_err()));
+
+            while let Some((event_id, event)) = query_stream.try_next().await? {
+                yield Ok((event_id, event));
+            }
+        })
+    }
+
+    async fn save_events(
+        &self,
+        _flow_id: &FlowID,
+        events: Vec<FlowEvent>,
+    ) -> Result<EventID, SaveEventsError> {
+        if events.is_empty() {
+            return Err(SaveEventsError::NothingToSave);
+        }
+
+        let mut tr = self.transaction.lock().await;
+
+        self.save_flow_updates_from_events(&mut tr, &events).await?;
+        let last_event_id = self.save_events_impl(&mut tr, &events).await?;
+
+        Ok(last_event_id)
+    }
+
+    async fn len(&self) -> Result<usize, InternalError> {
+        let mut tr = self.transaction.lock().await;
+        let connection_mut = tr.connection_mut().await?;
+
+        let result = sqlx::query!(
+            r#"
+            SELECT COUNT(event_id) as event_count
+                FROM flow_events
+            "#,
+        )
+        .fetch_one(connection_mut)
+        .await
+        .int_err()?;
+
+        let count = usize::try_from(result.event_count.unwrap()).int_err()?;
+        Ok(count)
+    }
+}
+
+////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
+
+#[async_trait::async_trait]
+impl FlowEventStore for PostgresFlowEventStore {
+    async fn new_flow_id(&self) -> Result<FlowID, InternalError> {
+        let mut tr = self.transaction.lock().await;
+        let connection_mut = tr.connection_mut().await?;
+
+        let result = sqlx::query!(
+            r#"
+            SELECT nextval('flow_id_seq') as new_flow_id
+            "#
+        )
+        .fetch_one(connection_mut)
+        .await
+        .int_err()?;
+
+        Ok(FlowID::new(result.new_flow_id.unwrap()))
+    }
+
+    async fn get_dataset_flow_run_stats(
+        &self,
+        dataset_id: &DatasetID,
+        flow_type: DatasetFlowType,
+    ) -> Result<FlowRunStats, InternalError> {
+        let mut tr = self.transaction.lock().await;
+
+        let connection_mut = tr.connection_mut().await?;
+        let maybe_attempt_result = sqlx::query!(
+            r#"
+            SELECT attempt.last_event_time as last_attempt_time
+            FROM (
+                SELECT e.event_id as event_id, e.event_time AS last_event_time
+                    FROM flow_events e, flows f
+                    WHERE
+                        e.event_type = 'FlowEventTaskFinished' AND
+                        f.dataset_id = $1 AND
+                        f.dataset_flow_type = $2 AND
+                        f.flow_id = e.flow_id
+                    ORDER BY e.event_id DESC
```

Comment:
The perfomance is fine, I've checked the query analyzer, we are hitting indices very well.

Reply:
I don't see what your suggestion is simplifying

Reply:
I don't see how this can run efficiently. At the minimum the query will need to:
- go through all "Finished" events of all datasets in reverse chronological order
- join all of them to `flows` table
- only then it can tell if this is the dataset_id / flow_type we're looking for.

So it may be using all the indexes to sort events and to join the events to `flows` table by `flow_id`, but I think it will still be processing a ton of unrelated data. And you are repeating this join pattern twice.

My suggestion finds one `flow_id` immediately using index and filters events only by that flow_id, reducing the search space dramatically. And the `flow_id` can be reused for the second event lookup.

If network RTT is a concern - I think all three steps can be combined in one query using CTE and `UNION ALL` to return both events.

Reply:
As discussed in Slack, the query planner is using 2 indices and does not show sign of any problem.
In the worst case, this run data is easily denormaliziable.